packages/base-contract/src/index.ts

```diff
@@ -100,17 +108,32 @@ export class BaseContract {
         }
         return txDataWithDefaults;
     }
-    protected static _throwIfRevertWithReasonCallResult(rawCallResult: string): void {
+    protected static _throwIfCallResultIsRevertError(rawCallResult: string): void {
         // Try to decode the call result as a revert error.
         let revert: RevertError;
         try {
-            revert = RevertError.decode(rawCallResult);
+            revert = decodeBytesAsRevertError(rawCallResult);
         } catch (err) {
             // Can't decode it as a revert error, so assume it didn't revert.
             return;
         }
         throw revert;
     }
+    protected static _throwIfThrownErrorIsRevertError(error: Error): void {
+        // Try to decode a thrown error.
+        let revertError: RevertError;
+        try {
+            revertError = decodeThrownErrorAsRevertError(error);
+            // Re-cast StringRevertErrors as plain Errors for backwards-compatibility.
+            if (revertError instanceof StringRevertError) {
+                throw new Error(revertError.values.message as string);
+            }
+        } catch (err) {
+            // Can't decode it.
+            return;
+        }
+        throw revertError;
+    }
```

Comment:
Woot! Thanks for doing this!

Reply:
Can you also use it in `awaitTransactionSuccessAsync` within the `tx.handlebars` abi-gen template?

Reply:
Yeah, I thought about this, but it would only work against ganache, so it would embed extra logic that's only meaningful in tests. Is there a way to generate/flag "debug" versions of contract wrappers?